### PR TITLE
add caching for pricing and instance type data

### PR DESCRIPTION
**Pull request by bwagner5**
Issue #, if available:
https://github.com/aws/amazon-ec2-instance-selector/issues/74

Description of changes:
 - Break up the ec2pricing pkg with two supporting structs, one for OD pricing and the other for SpotPricing
 - Implement Caching on OD Pricing, Spot Pricing, and InstanceTypes

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: cmd/main.go [hunk 1]
```diff
@@ -198,19 +212,35 @@ Full docs can be found at github.com/aws/amazon-` + binName
...
+				if err := instanceSelector.EC2Pricing.RefreshOnDemandCache(); err != nil {
+					log.Printf("There was a problem refreshing the on-demand pricing cache: %v", err)
+				}
+			}
 		}(wg)
 		go func(waitGroup *sync.WaitGroup) {
 			defer waitGroup.Done()
-			_ = instanceSelector.EC2Pricing.HydrateSpotCache(30)
+			if instanceSelector.EC2Pricing.SpotCacheCount() == 0 {
+				if err := instanceSelector.EC2Pricing.RefreshSpotCache(30); err != nil {
```
Comment (brycahta) [inline cmd/main.go:224]: `30` should be const if you're not planning to expose "days back" as a flag

File: pkg/ec2pricing/ec2pricing.go [hunk 2]
```diff
@@ -1,355 +1,115 @@
...
 const (
-	defaultSpotDaysBack = 30
-	productDescription  = "Linux/UNIX (Amazon VPC)"
-	serviceCode         = "AmazonEC2"
+	productDescription = "Linux/UNIX (Amazon VPC)"
+	serviceCode        = "AmazonEC2"
+)
+
+var (
+	DefaultSpotDaysBack = 30
```
Comment (brycahta) [inline pkg/ec2pricing/ec2pricing.go:32]: why change to var? Is it configurable?

File: pkg/ec2pricing/odpricing.go [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+		FullRefreshTTL: fullRefreshTTL,
+		DirectoryPath:  expandedDirPath,
+		pricingClient:  pricingClient,
+		cache:          cache.New(fullRefreshTTL, fullRefreshTTL),
+	}
+	if fullRefreshTTL <= 0 {
+		odPricing.Clear()
+		return odPricing
+	}
+	// Start the cache refresh job
```
Comment (brycahta) [inline pkg/ec2pricing/odpricing.go:71]: could you add why a goroutine vs. regular invocation in the comment? unless i'm missing something obvious..

File: pkg/instancetypes/instancetypes.go [hunk 4]
```diff
@@ -1,10 +1,168 @@
...
+
+func (p *Provider) Get(instanceTypes []string) ([]*Details, error) {
+	instanceTypeDetails := []*Details{}
+	describeInstanceTypeOpts := &ec2.DescribeInstanceTypesInput{}
+	if len(instanceTypes) != 0 {
+		for _, it := range instanceTypes {
+			if cachedIT, ok := p.cache.Get(it); ok {
+				instanceTypeDetails = append(instanceTypeDetails, cachedIT.(*Details))
+			} else {
+				// need to reassign so we're not sharing the loop iterators memory space
```
Comment (brycahta) [inline pkg/instancetypes/instancetypes.go:117]: nice 👍

**Review by brycahta — COMMENTED** [review]
Had a couple nits/questions, but awesome improvement overall and very clean

File: pkg/ec2pricing/ec2pricing.go (hunk 2 above)
Comment (bwagner5) [reply pkg/ec2pricing/ec2pricing.go:32]: Yes, if using `ec2pricing` as a library, the caller can configure the pkg by modifying the var.

File: pkg/ec2pricing/odpricing.go (hunk 3 above)
Comment (bwagner5) [reply pkg/ec2pricing/odpricing.go:71]: This is for a long running invocation of instance-selector, like a TUI, which might be coming soon...

File: cmd/main.go [hunk 5]
```diff
@@ -290,6 +320,7 @@ Full docs can be found at github.com/aws/amazon-` + binName
 	if itemsTruncated > 0 {
 		log.Printf("%d entries were truncated, increase --%s to see more", itemsTruncated, maxResults)
 	}
+	shutdown()
```
[the remark this answers is not in the document]
Comment (bwagner5) [reply cmd/main.go:308]: The reason I didn't defer is to avoid saving the cache on an error situation

**Review by brycahta — APPROVED** [review]
/lgtm 🚀

**Pull request merged**